LanguageFeatures/Patterns/type_inference_A19_t03.dart

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Calculate the static type of the pattern. Using that value type,
+/// recurse through the pattern again downwards to the leaf subpatterns filling
+/// in any missing types in the pattern. This process may also insert implicit
+/// coercions and casts from dynamic when values flow into a pattern during
+/// matching
+///
+/// @description Check that the calculation of the static type of a record
+/// pattern performs casts from dynamic and generic function instantiation
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns,records
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+
+T foo<T>(T t) => t;
+
+main() {
+  dynamic pi = 3.14;
+  final (double v1, n: double v2) = (pi, n: pi);
```

Comment:
Yes, this works! The record has a firm type (isn't affected by phase 1), and there should be dynamic checks because of the field types being `dynamic`.